src/empathy.1
.TH EMPATHY "1" "October 2007" "Telepathy project" "User Commands"
.SH NAME
empathy \- GNOME instant messaging client using Telepathy
.SH SYNOPSIS
empathy
.SH DESCRIPTION
Empathy consists of a rich set of reusable instant messaging widgets, and a
GNOME client using those widgets.
It uses Telepathy and Nokia's Mission Control, and reuses Gossip's UI.
.PP
The main user interface consists of a contact list window and an icon in the
notification area.
.SH OPTIONS
There are no command-line options.
.SH ENVIRONMENT
.TP
\fBEMPATHY_LOGFILE\fR=\fIfilename\fR
If set, debug output will go to the given file rather than to stderr.
.TP
\fBEMPATHY_DEBUG\fR=\fItype\fR
May be set to "all" for full debug output, or various undocumented options
(which may change from release to release) to filter the output.
.SH SEE ALSO
\fIhttp://telepathy.freedesktop.org/\fR, \fIhttp://live.gnome.org/Empathy\fR


NEWS
NEW in 0.23.1
==============
 - Show contact list when a chat becomes a chatroom. (Xavier Claessens).
 - Dispatch each Tube to the application handling its service. (Elliot Fairweather, Xavier Claessens).
 - Added introduction and started the Getting Started section of help manual (Milo Casagrande).
 - Add build system for help manual (Xavier Claessens).
 - Some code cleanup (Xavier Claessens).
 - Already open channels are considered OUTGOING. (Xavier Claessens).
 - Drop Chandler and Filter, do not use MC for dispatching channels, do it ourself. Everything is now in the same process. (Xavier Claessens).
 - Remove libtelepathy dep and use telepathy-glib instead. (Xavier Claessens).
 - Remove Room menu, chatrooms will be handled in a dedicated program. (Xavier Claessens).
 - Unify EmpathyPrivateChat and EmpathyGroupChat. (Xavier Claessens).
 - Unify EmpathyTpChatroom and EmpathyTpChat. (Xavier Claessens).
 - Do not autoconnect if networkmanager says we are not connected at startup. (Xavier Claessens).
 - Move modules that make no sense to be used in other applicaton from libempathy-gtk/ to src/ (Xavier Claessens).
 - Convert all files to glade3 (Laurent Bigonville).

Bugs fixed:
 - Fixed #526145, crash in Empathy Instant Messenger: Tryint to persuade Zdra,... (Xavier Claessens)
 - Fixed #475789, Compact mode + avatars = bad padding (Marco Barisione)
 - Fixed #479697, no sounds (Frederic Peters)
 - Fixed #489978, Removing NotHere applet from panel (Frederic Peters)
 - Fixed #496309, empathy translation issues (Andre Klapper)
 - Fixed #513676, Need a widget to configure AIM accounts (Andreas Henriksson)
 - Fixed #513679, Need a widget to configure SIP accounts (Baptiste Mille-Mathias)
 - Fixed #514380, Drop gnome-vfs dependency (Cosimo Cecchi)
 - Fixed #517776, Groupwise profile (Vincent Untz)
 - Fixed #519946, Confirmation dialog on remove contact/group (Bruno Dusausoy)
 - Fixed #521554, Diplay all the favourite rooms in the favourite room dialog (Marco Barisione)
 - Fixed #521570, Set to unsensitive the 'show avatars' option when using compact mode (Marco Barisione)
 - Fixed #522297, libempathy-gtk Python binding: cannot create a ContactListStore (Juan Pizarro)
 - Fixed #522528, Missing files in POTFILES.in (nshmyrev)
 - Fixed #522780, Couldn't build latest Empathy trunk (rev. 807) (Xavier Claessens)
 - Fixed #522862, Need a GUI to configure groupwise account (Vincent Untz)
 - Fixed #523703, Deprecated glib symbols used. (Frederic Peters)
 - Fixed #525039, Should not be started automatically via /etc/xdg/autostart (Frederic Peters)
 - Fixed #525084, Crash at GNOME startup (Xavier Claessens)
 - Fixed #525562, Empathy should not require python-config (Frederic Peters)
 - Fixed #526374, Crash in EmpathyContactListStore (Xavier Claessens)
 - Fixed #526927, "This call will be ended. Continue?" has no purpose when call status is disconnected (Frederic Peters)
 - Fixed #527452, Too difficult SIP configure UI (Frederic Peters)
 - Fixed #527454, Better call dialog (Naba Kumar)
 - Fixed #527713, crash when opening a chat with the keyboard (Frederic Peters)
 - Fixed #527876, "registrar" misspelled (Frederic Peters)
 - Fixed #527924, NotHere applet doesn't find icons (Frederic Peters)
 - Fixed #528229, chat dialogs way too large (Xavier Claessens)
 - Fixed #528287, Contact not shown in the contact list when he comes online (Xavier Claessens)
 - Fixed #528529, escaping text in strings using Pango markup (Frederic Peters)
 - Fixed #528542, Help is displayed on stderr, but stderr may go to log file (Frederic Peters)
 - Fixed #528596, Missing l10n for strings in call window (Frederic Peters)

Translations:
 - Updated es translation (Jorge Gonzalez).
 - Updated el translation (Giannis Katsampirhs).
 - Updated ca translation (Gil Forcada).
 - Updated Basque translation (Inaki Larranaga Murgoitio).

NEW in 0.22.0
==============
 - Queue received messages until the sender got his alias. (Xavier Claessens).

Bugs fixed:
 - Fixed #466521, Add an advanced expander in the generic setting widget (Marco Barisione)
 - Fixed #479534, remember conversation window sizes (Marco Barisione)
 - Fixed #501805, empathy should prevent to run multiple instances (Marco Barisione)
 - Fixed #505616, linkability of URL broken on _ (Marco Barisione)
 - Fixed #520788, Wrong button order in the favourite rooms dialog (Marco Barisione)
 - Fixed #521051, empathy_main_window_show() should use empathy_window_present() (Marco Barisione)

Translations:
 - Updated Finnish translation (ituohela).
 - Updated en_GB translation  (Philip Withnall).
 - Updated sv translation  (Daniel Nylander).
 - Updated it translation  (Milo Casagrande).
 - Updated Spanish translation  (Jorge Gonzalez).
 - Updated fr translation  (Robert-André Mauchin).

NEW in 0.21.91
==============
 - Blink the status icon for new incoming calls. (Xavier Claessens).
 - Close the media channel if stream engine crashes. (Xavier Claessens).
 - Change status to closed if stream engine dies or if a call fails. (Xavier Claessens).
 - Make use of generated code for using stream engine. (Xavier Claessens).
 - Add a GUI to configure IRC accounts (Guillaume Desmottes).
 - accounts-dialog: swap the create and previous button (Guillaume Desmottes).
 - save when adding/removing networks (Guillaume Desmottes).
 - check if network !=  NULL before trying to set its charset (Guillaume Desmottes).
 - New call code, should be more usable (Elliot Fairweather).
 - Try to create salut account if getting the gconf key fails. (Xavier Claessens).
 - Add a generic UI to make easy building UI for protocol specific settings. (Xavier Claessens).
 - Fix setting avatar on salut.  (Sjoerd Simons).
 - Use EMPATHY_SRCDIR environment variable to load uninstalled glade/xml/dtd files (Xavier Claessens).

Bugs fixed:
 - Fixed #505585, Improve .gitignore (Guillaume Desmottes)
 - Fixed #505622, add "check" framework support (Guillaume Desmottes)

src/empathy.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*- */
/*
 * Copyright (C) 2007-2008 Collabora Ltd.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 * 
 * Authors: Xavier Claessens <xclaesse@gmail.com>
 */

#include <config.h>

#include <stdlib.h>
#include <errno.h>

#include <glib.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <gdk/gdkx.h>

#include <libebook/e-book.h>

#include <telepathy-glib/util.h>
#include <libmissioncontrol/mc-account.h>
#include <libmissioncontrol/mission-control.h>

#include <libempathy/empathy-idle.h>
#include <libempathy/empathy-utils.h>
#include <libempathy/empathy-debug.h>

#include <libempathy-gtk/empathy-conf.h>

#include "empathy-main-window.h"
#include "empathy-status-icon.h"
#include "bacon-message-connection.h"

#define DEBUG_DOMAIN "EmpathyMain"

static BaconMessageConnection *connection = NULL;

static void
service_ended_cb (MissionControl *mc,
		  gpointer        user_data)
{
	empathy_debug (DEBUG_DOMAIN, "Mission Control stopped");
}

static void
operation_error_cb (MissionControl *mc,
		    guint           operation_id,
		    guint           error_code,
		    gpointer        user_data)
{
	const gchar *message;

	switch (error_code) {
	case MC_DISCONNECTED_ERROR:
		message = _("Disconnected");
		break;
	case MC_INVALID_HANDLE_ERROR:
		message = _("Invalid handle");
		break;
	case MC_NO_MATCHING_CONNECTION_ERROR:
		message = _("No matching connection");
		break;
	case MC_INVALID_ACCOUNT_ERROR:
		message = _("Invalid account");
		break;
	case MC_PRESENCE_FAILURE_ERROR:
		message = _("Presence failure");
		break;
	case MC_NO_ACCOUNTS_ERROR:
		message = _("No accounts");
		break;
	case MC_NETWORK_ERROR:
		message = _("Network error");
		break;
	case MC_CONTACT_DOES_NOT_SUPPORT_VOICE_ERROR:
		message = _("Contact does not support voice");
		break;
	case MC_LOWMEM_ERROR:
		message = _("Lowmem");
		break;
	case MC_CHANNEL_REQUEST_GENERIC_ERROR:
		message = _("Channel request generic error");
		break;
	case MC_CHANNEL_BANNED_ERROR:
		message = _("Channel banned");
		break;
	case MC_CHANNEL_FULL_ERROR:
		message = _("Channel full");
		break;
	case MC_CHANNEL_INVITE_ONLY_ERROR:
		message = _("Channel invite only");
		break;
	default:
		message = _("Unknown error code");
	}

	empathy_debug (DEBUG_DOMAIN, "Error during operation %d: %s",
		       operation_id, message);
}

static void
use_nm_notify_cb (EmpathyConf *conf,
		  const gchar *key,
		  gpointer     user_data)
{
	EmpathyIdle *idle = user_data;
	gboolean     use_nm;

	if (empathy_conf_get_bool (conf, key, &use_nm)) {
		empathy_idle_set_use_nm (idle, use_nm);
	}
}

static void
create_salut_account (void)
{
	McProfile  *profile;
	McProtocol *protocol;
	gboolean    salut_created = FALSE;
	McAccount  *account;
	GList      *accounts;
	EBook      *book;
	EContact   *contact;
	gchar      *nickname = NULL;
	gchar      *first_name = NULL;
	gchar      *last_name = NULL;
	gchar      *email = NULL;
	gchar      *jid = NULL;
	GError     *error = NULL;

	/* Check if we already created a salut account */
	empathy_conf_get_bool (empathy_conf_get(),
			       EMPATHY_PREFS_SALUT_ACCOUNT_CREATED,
			       &salut_created);
	if (salut_created) {
		return;
	}

	empathy_debug (DEBUG_DOMAIN, "Try to add a salut account...");

	/* Check if the salut CM is installed */
	profile = mc_profile_lookup ("salut");
	protocol = mc_profile_get_protocol (profile);
	if (!protocol) {
		empathy_debug (DEBUG_DOMAIN, "Salut not installed");
		g_object_unref (profile);
		return;
	}
	g_object_unref (protocol);

	/* Get self EContact from EDS */
	if (!e_book_get_self (&contact, &book, &error)) {
		empathy_debug (DEBUG_DOMAIN, "Failed to get self econtact: %s",
			       error ? error->message : "No error given");
		g_clear_error (&error);
		g_object_unref (profile);
		return;
	}

	empathy_conf_set_bool (empathy_conf_get (),
			       EMPATHY_PREFS_SALUT_ACCOUNT_CREATED,
			       TRUE);

	/* Check if there is already a salut account */
	accounts = mc_accounts_list_by_profile (profile);
	if (accounts) {
		empathy_debug (DEBUG_DOMAIN, "There is already a salut account");
		mc_accounts_list_free (accounts);
		g_object_unref (profile);
		return;
	}

	account = mc_account_create (profile);
	mc_account_set_display_name (account, _("People nearby"));
	
	nickname = e_contact_get (contact, E_CONTACT_NICKNAME);
	first_name = e_contact_get (contact, E_CONTACT_GIVEN_NAME);
	last_name = e_contact_get (contact, E_CONTACT_FAMILY_NAME);
	email = e_contact_get (contact, E_CONTACT_EMAIL_1);
	jid = e_contact_get (contact, E_CONTACT_IM_JABBER_HOME_1);
	
	if (!tp_strdiff (nickname, "nickname")) {
		g_free (nickname);
		nickname = NULL;
	}

	empathy_debug (DEBUG_DOMAIN, "Salut account created:\n"
				     "  nickname=%s\n"
				     "  first-name=%s\n"
				     "  last-name=%s\n"
				     "  email=%s\n"
				     "  jid=%s\n",
		       nickname, first_name, last_name, email, jid);

	mc_account_set_param_string (account, "nickname", nickname ? nickname : "");
	mc_account_set_param_string (account, "first-name", first_name ? first_name : "");
	mc_account_set_param_string (account, "last-name", last_name ? last_name : "");
	mc_account_set_param_string (account, "email", email ? email : "");
	mc_account_set_param_string (account, "jid", jid ? jid : "");

	g_free (nickname);
	g_free (first_name);
	g_free (last_name);
	g_free (email);
	g_free (jid);
	g_object_unref (account);
	g_object_unref (profile);
	g_object_unref (contact);
	g_object_unref (book);
}

/* The code that handles single-instance and startup notification is
 * copied from gedit.
 *
 * Copyright (C) 2005 - Paolo Maggi 
 */
static void
on_bacon_message_received (const char *message,
			   gpointer    data)
{
	GtkWidget *window = data;
	guint32    startup_timestamp;

	g_return_if_fail (message != NULL);

	empathy_debug (DEBUG_DOMAIN,
		       "Other instance launched, presenting the main window "
		       "(message is '%s')", message);

	startup_timestamp = atoi (message);

	/* Set the proper interaction time on the window.
	 * Fall back to roundtripping to the X server when we
	 * don't have the timestamp, e.g. when launched from
	 * terminal. We also need to make sure that the window
	 * has been realized otherwise it will not work. lame. */
	if (startup_timestamp == 0) {
		/* Work if launched from the terminal */
		empathy_debug (DEBUG_DOMAIN, "Using X server timestamp as a fallback");

		if (!GTK_WIDGET_REALIZED (window)) {
			gtk_widget_realize (GTK_WIDGET (window));
		}

		startup_timestamp = gdk_x11_get_server_time (window->window);
	}

	gtk_window_present_with_time (GTK_WINDOW (window), startup_timestamp);
}

static guint32
get_startup_timestamp ()
{
	const gchar *startup_id_env;
	gchar       *startup_id = NULL;
	gchar       *time_str;
	gchar       *end;
	gulong       retval = 0;

	/* we don't unset the env, since startup-notification
	 * may still need it */
	startup_id_env = g_getenv ("DESKTOP_STARTUP_ID");
	if (startup_id_env == NULL) {
		goto out;
	}

	startup_id = g_strdup (startup_id_env);

	time_str = g_strrstr (startup_id, "_TIME");
	if (time_str == NULL) {
		goto out;
	}

	errno = 0;

	/* Skip past the "_TIME" part */
	time_str += 5;

	retval = strtoul (time_str, &end, 0);
	if (end == time_str || errno != 0)
		retval = 0;

 out:
	g_free (startup_id);

	return (retval > 0) ? retval : 0;
}

int
main (int argc, char *argv[])
{
	guint32            startup_timestamp;
	EmpathyStatusIcon *icon;
	GtkWidget         *window;
	MissionControl    *mc;
	EmpathyIdle       *idle;
	gboolean           autoconnect = TRUE;
	gboolean           no_connect = FALSE; 
	GError            *error = NULL;
	GOptionEntry       options[] = {
		{ "no-connect", 'n',
		  0, G_OPTION_ARG_NONE, &no_connect,
		  N_("Don't connect on startup"),
		  NULL },
		{ NULL }
	};

	bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	startup_timestamp = get_startup_timestamp ();

	if (!gtk_init_with_args (&argc, &argv,
				 _("- Empathy Instant Messenger"),
				 options, GETTEXT_PACKAGE, &error)) {
		empathy_debug (DEBUG_DOMAIN, error->message);
		return EXIT_FAILURE;
	}

	empathy_debug_set_log_file_from_env ();

	g_set_application_name (PACKAGE_NAME);

	gtk_window_set_default_icon_name ("empathy");
	gtk_icon_theme_append_search_path (gtk_icon_theme_get_default (),
					   PKGDATADIR G_DIR_SEPARATOR_S "icons");

        /* Setting up the bacon connection */
	connection = bacon_message_connection_new ("empathy");
	if (connection != NULL) {
		if (!bacon_message_connection_get_is_server (connection)) {
			gchar *message;

			empathy_debug (DEBUG_DOMAIN, "Activating existing instance");

			message = g_strdup_printf ("%" G_GUINT32_FORMAT,
						   startup_timestamp);
			bacon_message_connection_send (connection, message);

			/* We never popup a window, so tell startup-notification
			 * that we are done. */
			gdk_notify_startup_complete ();

			g_free (message);
			bacon_message_connection_free (connection);

			return EXIT_SUCCESS;
		}
	} else {
		g_warning ("Cannot create the 'empathy' bacon connection.");
	}

	/* Setting up MC */
	mc = empathy_mission_control_new ();
	g_signal_connect (mc, "ServiceEnded",
			  G_CALLBACK (service_ended_cb),
			  NULL);
	g_signal_connect (mc, "Error",
			  G_CALLBACK (operation_error_cb),
			  NULL);

	/* Setting up Idle */
	idle = empathy_idle_new ();
	empathy_idle_set_auto_away (idle, TRUE);
	use_nm_notify_cb (empathy_conf_get (), EMPATHY_PREFS_USE_NM, idle);
	empathy_conf_notify_add (empathy_conf_get (), EMPATHY_PREFS_USE_NM,
				 use_nm_notify_cb, idle);

	/* Autoconnect */
	empathy_conf_get_bool (empathy_conf_get(),
			       EMPATHY_PREFS_AUTOCONNECT,
			       &autoconnect);
	if (autoconnect && ! no_connect &&
	    empathy_idle_get_state (idle) <= MC_PRESENCE_OFFLINE) {
		empathy_idle_set_state (idle, MC_PRESENCE_AVAILABLE);
	}
	
	create_salut_account ();

	/* Setting up UI */
	window = empathy_main_window_show ();
	icon = empathy_status_icon_new (GTK_WINDOW (window));

	if (connection) {
		/* We se the callback here because we need window */
		bacon_message_connection_set_callback (connection,
						       on_bacon_message_received,
						       window);
	}

	gtk_main ();

	empathy_idle_set_state (idle, MC_PRESENCE_OFFLINE);

	g_object_unref (mc);
	g_object_unref (idle);
	g_object_unref (icon);

	return EXIT_SUCCESS;
}

